Add pin feature for messages

jrainville opened the pull request:
Fixes #1905 

Put as draft because it is waiting for the status-go PR (https://github.com/status-im/status-go/pull/2180)  to be merged and also because it is missing the status-go signal. Also, startWallet was removed from status-go, so I commented it for now, but I'll change it once Richard has a PR.

Adds the Pin and Unpin feature to non-public channels.

Adds a popup to see the pinned messages.

They also work for communities

Missing stuff:
- The API doesn't give the person who pinned the message, so I couldn't follow the design on this
- There is currently no signal when a pin or unpin is made, so it doesn't update automatically.
- Changes to startWallet because of the breaking change in status-go


![image](https://user-images.githubusercontent.com/11926403/115296222-b3120b00-a128-11eb-8ed9-12b93af20d48.png)

![image](https://user-images.githubusercontent.com/11926403/115296469-ff5d4b00-a128-11eb-94af-452f4160d621.png)

jrainville commented on src/status/libstatus/wallet.nim:
```diff
@@ -84,7 +84,8 @@ proc validateMnemonic*(mnemonic: string): string =
 proc startWallet*(watchNewBlocks: bool) =
-  discard status_go.startWallet(watchNewBlocks) # TODO: true  to watch trx
+  discard
+  # discard status_go.startWallet(watchNewBlocks) # TODO: true  to watch trx
```
This is no longer in status-go. It was a breaking change in the latest status-go. Richard is working on a fix, but it was easier for me for now to just comment it.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 58859f9f | [#1](https://ci.status.im/job/status-desktop/job/branches/job/macos/job/PR-2291/1/) | 2021-04-19 20:14:39 | ~8 min | `macos` | [:package:`dmg`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-210419-200635-58859f-pr.dmg) |

jrainville commented on src/app/chat/views/message_list.nim:
```diff
@@ -231,9 +243,11 @@ QtObject:
   proc add*(self: ChatMessageList, message: Message) =
     if self.messageIndex.hasKey(message.id): return # duplicated msg
 
+    debug "New message", chatId = self.id, id = message.id, text = message.text
```
Commenting here to rembeber to remove this

anastasiyaig commented:
1. 3 buttons option is shown for **public** chats with no actions, when clicking it -> i cant close the menu, it will be shown forever

Possible expected result: remove 3 dots from public channels

![Screenshot 2021-05-19 at 15.31.54.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/8cd1d577-001e-4e1b-9fa3-f011b6ed017d)

2. Pin icon is too small 

![Screenshot 2021-05-19 at 16.35.35.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/614ce63f-d4a1-41c9-8d38-d6ea378be27d)

3. Pinned **by** is missing (known limitation )

4. Message(s) does not reflect the count of messages, so it is always Message and not a plural. Also the representation of list of pinned messages is in yellow and quite messy
![Screenshot 2021-05-19 at 16.45.00.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/0f57deca-54b1-4778-b01a-eea84a8f70c3)

5. no reference to Pinned messages in Group chat profile

![Screenshot 2021-05-19 at 16.47.38.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/a5c2fabf-5595-469b-8440-9958a92b1eab)

6. context menu is glued to a particular part of the screen, so resizing the window breaks it from the 3 dots
![Screenshot 2021-05-19 at 16.56.19.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/433c5a74-f79a-4065-87f5-606193955037)

7. wrong hover on Pin / Unpin
![Screenshot 2021-05-19 at 16.59.53.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/c2649081-bdfa-4391-9027-6af01d76401c)

simonricoo commented:
7. Is just not updated design from the past, so the hover is ok :) It came with the latest redesign. However the length of the dropdown is quite long and **Pin** and **Copy** icons are blurred :((((((

simonricoo commented:
![115296469-ff5d4b00-a128-11eb-94af-452f4160d621.png](https://images.zenhubusercontent.com/5eba533dc999132939d1d6ae/141ee52b-171c-42c8-9cae-c011724f75f4)
![Screen_Shot_2021-05-19_at_16_55_54.png](https://images.zenhubusercontent.com/5eba533dc999132939d1d6ae/a83c8659-b2e7-4633-a36e-47b5b858bc55)

jrainville commented:
> Possible expected result: remove 3 dots from public channels, add possibility to close context menu by clicking out from the menu area

I know this was not part of the design, but I find that it makes sense to be able to see the other options using the three dots. So being able to copy the text from there as well as the right-click. Anyway, I do hide the "Pin" button when it's a public chat

jrainville commented:
@anastasiyaig I fixed the issues you reported. Thanks for the thorough testing.

Some of them I couldn't fix or I didn't for various reasons, here they are:
- I added the `Pinned by` part, BUT, there is an issue in Status-go where if you close the app and reopen, the pinned by now shows the author of the message. I opened the issue here: https://github.com/status-im/status-go/issues/2234
- the point 6, where the menu is glued to the part of the screen should be opened in another issue, since it's something that affect all popup menus and not just the pin one.
- like Simon said, 7 is just the design not being up to date

@simonricoo sadly, I can't fix the rounded corner problem. The workaround we usually use to have different corners doesn't work if the color is translucid, because we can see the other rectangle underneath the first one.

anastasiyaig commented:
@jrainville awesome work , thanks for putting an effort to this!

i logged the menu issue separately https://github.com/status-im/status-desktop/issues/2584, also thanks for opening the issue on go side! I checked -> it happens for communities channels and group chats, but for 1 x1 is okay 

There are small points which we probably can fix easily:

- increase click area for this menu (3 dots) -> for now there are places which highlight the button (meaning it is active) but clicking does not make any effect

![Screenshot 2021-05-21 at 11.54.41.png](https://images.zenhubusercontent.com/6074560cfd1d342d606f37d3/913b45b3-5436-40bf-87ff-844a093ae0ec)

- menu (3 dots) overlaps with pinned sticker message

Also, due to https://github.com/status-im/status/issues/11#issuecomment-773982069 unpin should be done:

1. by any admin of group chat and in this PR anyone can unpin no matter if he is admin or not
2. by admin in community channels and in this PR anyone can unpin no matter if he is admin or not

Can we please fix that as well?

jrainville commented:
@anastasiyaig thanks again for the testing.

I fixed the overlap issue and also made it so that only admins can pin in communities and group chats

anastasiyaig commented:
@jrainville 

1. The clicking area is still small with these 3 dots , so i still need to find exact place where to click it to get the options (please pay attention to the Pinned messages pop-up)

<img width="1552" alt="Screenshot 2021-05-24 at 12 45 55" src="https://user-images.githubusercontent.com/82375995/119329584-0cbca680-bc8e-11eb-8ee8-27899333cbd8.png">


2. Pinned message is still overlapping the other message

3. Pinned messages are not shown for non-admin **in a group chat**, it is kind of regression since it was there in previous build

<img width="1430" alt="Screenshot 2021-05-24 at 12 24 36" src="https://user-images.githubusercontent.com/82375995/119328135-75a31f00-bc8c-11eb-8ecf-98f8edf7e57e.png">

4. Group chat -> profile -> pinned messages throws an error
`qrc:/app/AppLayouts/Chat/components/GroupInfoPopup.qml:214: ReferenceError: pinnedMessagesPopupComponent is not defined`
5. Group chat -> profile -> pinned messages misses the icon
<img width="1430" alt="Screenshot 2021-05-24 at 12 31 34" src="https://user-images.githubusercontent.com/82375995/119327829-12b18800-bc8c-11eb-8838-219d111f4375.png">
6. Communities -> manage community: no pinned messages view option at all https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?node-id=5971%3A42716

![image](https://user-images.githubusercontent.com/82375995/119331051-94ef7b80-bc8f-11eb-8519-ca47c4a4fbad.png)

7. also with point 6 we need zero screen ( 0 pinned messages)
![image](https://user-images.githubusercontent.com/82375995/119331208-c700dd80-bc8f-11eb-956a-ee76ee7c7e4c.png)

richard-ramos reviewed (approved):
very nice!

the only change it requires is to update the status-go version it is using.

jrainville commented:
@anastasiyaig I fixed the issues again.

Details about some of them:

1. I couldn't reproduce this one on Linux nor on Windows. Could you show a video? Does it only happen in the first element of the list or all of them?

2. this one is very hard to fix. I fixed it last time, but apparently my fix only works when you reopen the app. If you send a new message, then the fix doesn't work anymore (it doesn't calculate the Z index of the new element correctly for some reason). Since I couldn't fix it easily today, I opened a new issue to fix it later since I don't think it should be a blocker to merge the PR https://github.com/status-im/status-desktop/issues/2598

3. This is not an issue with this PR, it is either an issue in status-go or desktop as a whole, we'd need to check. Basically, as you can see on the right screen of your screenshot, the pinned messages are **before** the lest account joined the group chat. So the left account doesn't see (or doesn't have access?) to the messages. That's why it doesn't have the pinned messages buttons. If you add a new message and pin a new message after the left account joined, you'd see the button then. It's probably an issue of desktop where we don't fetch past messages when joining a group

6. The button is in the community channel edit popup, which is not merged yet. I added an issue to add the button later: https://github.com/status-im/status-desktop/issues/2597

jrainville commented:
Rebased and updated status-go to match master

anastasiyaig commented:
lets merge the PR and address the issues with backend separately
